src/main/java/org/testng/providers/IObjectProvider.java

```diff
@@ -0,0 +1,29 @@
+package org.testng.providers;
+
+import java.lang.annotation.Annotation;
+import java.util.Map;
+import org.testng.IClass;
+import org.testng.ITestObjectFactory;
+import org.testng.internal.annotations.IAnnotationFinder;
+import org.testng.xml.XmlTest;
+
+public interface IObjectProvider {
+
+  Object create(Class<?> declaringClass,
+      Map<Class<?>, IClass> classes,
+      XmlTest xmlTest,
+      IAnnotationFinder finder,
+      ITestObjectFactory factory,
+      boolean create,
+      String errorMsgPrefix);
+
+  Object create(IClass iClass, Class<?> clazz);
+
+  default Object create(Class<?> clazz) {
+    return null;
+  }
+
+  boolean supportsDependencyInjection();
+
+  Class<? extends Annotation> associatedAnnotation();
```

Comment:
I have some doubts about this method: An object provider could support many annotations and an annotation could be supported by many providers.

Reply:
@juherr - I put this in to solve a way of being able to uniquely identify an implementation rather than anything else. That is why I put in a new annotation. So its the implementation's way of telling TestNG what is its unique name

Reply:
`String getName()` will be valid too.
In fact, I don't see a strong relationship between the object provider and an annotation even if an object provider can support a specific annotation.

Reply:
Choosing a string means that i need to ensure that we have unique values all across. Using an annotation automatically guarantees me that because its hard to intentionally build colliding annotations with the same fully qualified package name. I can even flip that to just using the class name (using `getClass()`). I dont have any specific preference on this.